<!DOCTYPE html><html lang="en"><head><script async src="https://www.googletagmanager.com/gtag/js?id=UA-114897551-4"></script><script>window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'UA-114897551-4');
</script><script type="text/javascript" src="//platform-api.sharethis.com/js/sharethis.js#property=5ac2443d1fff98001395ab6c&amp;product=sticky-share-buttons" async="async"></script><title>Dealing With Exceptional Behavior - Intro to Computer Science | Coder Coacher - Coaching Coders</title><meta content="Dealing With Exceptional Behavior - Intro to Computer Science - All technical stuff in one place" name="description"><meta name="keywords" content="education, coding, programming, technology, nodejs, mongodb, software, computer science, engineering, teaching, coaching, coder, learning, java, kotlin, machine learning, AI, ML, tech talks, angular, javascript, js, typescript"><meta name="viewport" content="width=device-width, initial-scale=1.0"><link rel="stylesheet" href="/css/font.css"><link rel="stylesheet" href="/css/bootstrap.css"><link rel="stylesheet" href="/css/style.css"><link rel="stylesheet" href="/css/coder-coacher.css"></head><body><div class="container-fluid"><h1 class="site-title"><a href="/">Coder Coacher</a></h1><hr><h4 class="site-subtitle text-right">Coaching Coders</h4></div><div id="amzn-assoc-ad-99d6751e-2392-4004-ad16-73aa8385d9d0"></div><script async src="//z-na.amazon-adsystem.com/widgets/onejs?MarketPlace=US&amp;adInstanceId=99d6751e-2392-4004-ad16-73aa8385d9d0"></script><div class="post__breadcrumb"><div class="container"><ol class="breadcrumb"><li><a href="/">Coder Coacher</a></li><li><a href="/Udacity/">Udacity</a></li><li class="active">â¤µ</li></ol></div></div><h2 class="post__title"><b>Dealing With Exceptional Behavior - Intro to Computer Science</b></h2><h5 class="post__date">2015-02-23</h5><div class="container"><div class="video-responsive"><iframe width="560" height="315" src="https://www.youtube.com/embed/KN91SMT2SlM" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe></div><div class="post__text">in the real world bad things do happen
like files not being writable or dis
painful and we don't want our programs
just to crash when that happens we want
to deal with it many programming
languages including Python provide a way
of doing this and it's called an
exception here's an example what's going
on I'm going to open the file called
blimpy to read from the current
directory and when i do this i get an
error the file didn't exist and there i
got is called an i/o error and it gives
me a message this is really a crash I
did this in the Python interpreter but
if this happened inside a program well
that would end execution and we'd be
done if we want our programs to work or
at least fail gracefully in the real
world where things like this can happen
instead of allowing this failure to
happen what we need to do is catch it
there's no way for open to return a file
object because we asked for a file that
doesn't exist and that can't be read so
there's no way to do this assignment we
can imagine a different interface to
open where it would return null or some
special object and then we'd have to
test every time we did it open to see if
it worked since we assume it normally
works we don't want to have to do that
test all the time instead we want to
deal with failures like this as
exceptional cases so here's what
happened in the example we're calling
the open operation which leads to call
in the file module normally that would
return a file instance which would be
assigned to the variable F in this case
something went wrong there's no file
that can be returned instead what
happens is an exception will be throwing
this isn't a normal return so execution
doesn't continue as it would with the
assignment it continues somewhere else
instead of being returned normally this
exception doesn't continue execution
normally it gets caught by an exception
handler if there is one they can have
more than five fingers if there's no
exception handler it causes the program
to crash and that's what we saw in the
interpreter and I'm sure you've seen
programs that end with that type of
crash alrdy how do we make these
handlers to try to prevent the program
from crashing to do that we need to try
catch block here's the syntax we use try
to introduce a block anything that
happens in that block if it raises an
exception it will jump to the exception
handlers if there's a handler that
matches the exception that was raised
and we'll talk more about how that
matching happens then the exception
handler will run so we can jump from
anywhere in the middle of this block
wherever an exception is raised to the
handler that handles that exception</div></div><div class="container-fluid bottom-ad"><div id="amzn-assoc-ad-6a809dda-347a-4187-8a86-91faf94575da"></div><script async src="//z-na.amazon-adsystem.com/widgets/onejs?MarketPlace=US&amp;adInstanceId=6a809dda-347a-4187-8a86-91faf94575da"></script></div><div class="text-center">We are a participant in the Amazon Services LLC Associates Program, an affiliate advertising program designed to provide a means for us to earn fees by linking to Amazon.com and affiliated sites.</div><script>(function(w, d){
    var b = d.getElementsByTagName('body')[0];
    var s = d.createElement("script"); s.async = true;
    var v = !("IntersectionObserver" in w) ? "8.6.0" : "10.4.2";
    s.src = "https://cdnjs.cloudflare.com/ajax/libs/vanilla-lazyload/" + v + "/lazyload.min.js";
    w.lazyLoadOptions = {};
    b.appendChild(s);
}(window, document));</script></body></html>